### PR TITLE
refactor: Context handling updates

### PR DESCRIPTION
**Pull request by james-milligan**
Changes to introduce proper context handling for graceful shutdown

https://github.com/open-feature/flagd/issues/128

### CODE REVIEW

File: pkg/runtime/runtime.go [hunk 1]
```diff
@@ -39,29 +39,27 @@ func startSyncer(ctx context.Context, notifier chan sync.INotify, syncr sync.ISy
...
-				case sync.DefaultEventTypeModify:
-					logger.Info("Configuration modified")
-					if err := updateState(ctx, syncr); err != nil {
-						log.Error(err)
-					}
-				case sync.DefaultEventTypeDelete:
-					logger.Info("Configuration deleted")
-				case sync.DefaultEventTypeReady:
-					logger.Info("Notifier ready")
+			w := <-notifier
```
Comment (skyerus) [inline pkg/runtime/runtime.go:42]: What's the reason for getting rid of the `<-ctx.Done()` cleanup?

File: pkg/sync/http_sync.go [hunk 2]
```diff
@@ -51,7 +51,7 @@ func (fs *HTTPSync) fetchBodyFromURL(ctx context.Context, url string) ([]byte, e
 	}
 	defer func() { _ = resp.Body.Close() }()
 
-	body, err := ioutil.ReadAll(resp.Body)
+	body, err := io.ReadAll(resp.Body)
```
Comment (AlexsJones) [inline pkg/sync/http_sync.go:54]: What is the benefit/difference here?

File: pkg/service/http_service.go [hunk 3]
```diff
@@ -111,14 +115,31 @@ func (s *HTTPService) Serve(ctx context.Context, eval eval.IEvaluator) error {
...
 
-	go func() { handleServiceError(httpServer.Serve(httpl)) }() // HTTP
-
-	go func() { handleServiceError(httpServer.Serve(httpsl)) }() // HTTPS
-
-	go func() { handleServiceError(grpcServer.Serve(gorpcl)) }() // GRPC
-	go func() { handleServiceError(tlsm.Serve()) }()
-	go func() { handleServiceError(tcpm.Serve()) }()
-	<-ctx.Done()
+	g.Go(func() error {
```
Comment (AlexsJones) [inline pkg/service/http_service.go:118]: How do errors work here? 
If the first routine exits immediately does the error return get waited on until all return?
Or does the exit cause all other routines to exit?

File: pkg/runtime/runtime.go (hunk 1 above)
Comment (james-milligan) [reply pkg/runtime/runtime.go:42]: Following our discussion I'm going to roll this back

File: pkg/service/http_service.go (hunk 3 above)
Comment (james-milligan) [reply pkg/service/http_service.go:118]: any error returned by one of the goroutines in the `errgroup` will cancel the group context will cancel the group context and `<-gCtx.Done()` will no longer block. The original error will be passed out from `g.Wait()`

**Comment by james-milligan** [comment]
closed as changes are required for [this PR](https://github.com/open-feature/flagd/pull/135), if those changes are rejected I will reopen this PR

**Pull request closed without merging**